File: 3776.minimum-moves-to-balance-circular-array.py
#
# @lc app=leetcode id=3776 lang=python3
#
# [3776] Minimum Moves to Balance Circular Array
#

# @lc code=start
class Solution:
    def minMoves(self, balance: List[int]) -> int:
        if sum(balance) < 0:
            return -1

        n = len(balance)
        moves = 0
        neg_index = next((i for i in range(n) if balance[i] < 0), None)
        if neg_index is None:
            return 0  # No negative balance means no moves needed.
        
        while balance[neg_index] < 0:
            for i in range(n):
                if balance[i] > 0:
                    transfer_amount = min(-balance[neg_index], balance[i])
                    balance[i] -= transfer_amount
                    balance[neg_index] += transfer_amount
                    moves += transfer_amount
                if balance[neg_index] >= 0:
                    break
            neg_index = (neg_index + 1) % n # Circular array consideration.
        
        return moves # @lc code=end

File: 3776.minimum-moves-to-balance-circular-array.cpp
#
# @lc app=leetcode id=3776 lang=cpp
#
# [3776] Minimum Moves to Balance Circular Array
#

# @lc code=start
#include <vector>
#include <algorithm>
using namespace std;
class Solution {
public:
    long long minMoves(vector<int>& balance) {
        long long total = 0; // To store total moves needed.
        int n = balance.size();
        int negIndex = -1; // Index of negative balance if any.
        for (int i = 0; i < n; ++i) {
            if (balance[i] < 0) negIndex = i; // Identify negative balance index.
            total += abs(balance[i]); // Calculate total absolute balances.
        }
        if (negIndex == -1) return 0; // No need to move if no negatives.
        vector<long long> prefix(n+1, 0); 
        for (int i = 1; i <= n; ++i) prefix[i] = prefix[i-1] + balance[i-1]; // Prefix sum calculation.
        auto check = [&]() -> bool { 
            for (int i = 0; i < n; ++i) { 
                if (prefix[i] - prefix[negIndex] < 0 || prefix[n] - prefix[negIndex] + prefix[i] < 0) return false; 
            } 
            return true; 
        }; 
        return check() ? total : -1; // Return total moves or -1 if impossible.
    }
};
# @lc code=end